File: src/models/irt.ts
import type { ResponseEvent } from '../core/event-schema.js';
import type {
  FittedModel,
  MasteryLearnerEntry,
  MasteryReport,
  ScoringModel,
} from '../core/scoring-model.js';

export interface IrtItemParams {
  skillId: string;
  /** Discrimination. Higher means the item separates high/low ability learners more sharply. */
  a: number;
  /** Difficulty, on the same scale as theta. Higher means a harder skill. */
  b: number;
}

export interface IrtLearnerResult {
  learnerId: string;
  /** Estimated ability. */
  theta: number;
  responseCount: number;
}

export interface IrtConfig {
  /** Number of gradient-ascent iterations to run. */
  iterations?: number;
  /** Learning rate for the gradient-ascent updates. */
  learningRate?: number;
  /**
   * L2 regularization strength pulling theta and b toward 0 and a toward 1.
   * This is what keeps the joint MLE finite for learners/items with a
   * perfect (all-correct or all-incorrect) response record, where the
   * unregularized likelihood is maximized at +/-infinity.
   */
  regularization?: number;
}

const DEFAULT_IRT_CONFIG: Required<IrtConfig> = {
  iterations: 500,
  learningRate: 0.5,
  regularization: 0.01,
};

export interface IrtFittedModel extends FittedModel {
  modelName: 'irt';
  items: IrtItemParams[];
  learners: IrtLearnerResult[];
  /** Raw (learner, item, correct) triples the fit converged against, for score() to attach predictions. */
  responses: { learnerId: string; skillId: string; correct: boolean }[];
}

/** Numerically stable logistic sigmoid. */
function sigmoid(z: number): number {
  if (z >= 0) {
    const e = Math.exp(-z);
    return 1 / (1 + e);
  }
  const e = Math.exp(z);
  return e / (1 + e);
}

/** P(correct) under the 2-parameter logistic model. */
export function probabilityCorrect(theta: number, a: number, b: number): number {
  return sigmoid(a * (theta - b));
}

interface JmleResult {
  theta: Map<string, number>;
  a: Map<string, number>;
  b: Map<string, number>;
}

/**
 * Joint maximum-likelihood estimation for the 2PL IRT model via batch
 * gradient ascent on the log-likelihood, with a small L2 prior (regularizing
 * theta/b toward 0 and a toward 1) that keeps estimates finite for learners
 * or skills with an all-correct or all-incorrect record.
 *
 * The 2PL model is only identified up to an additive shift and multiplicative
 * scale of theta (z = a*(theta-b) is unchanged by shifting theta and b by the
 * same constant, or by scaling theta/b by s while dividing a by s). To pin
 * down a single solution, after every iteration the theta distribution is
 * re-centered to mean 0 and re-scaled to standard deviation 1, applying the
 * matching inverse transform to b and a so every predicted probability is
 * left exactly unchanged. This is the standard way JMLE implementations fix
 * the person-parameter scale.
 */
function fitJmle(
  learnerIds: string[],
  itemIds: string[],
  responses: { learnerIndex: number; itemIndex: number; correct: boolean }[],
  config: Required<IrtConfig>,
): JmleResult {
  const theta = new Float64Array(learnerIds.length).fill(0);
  const a = new Float64Array(itemIds.length).fill(1);
  const b = new Float64Array(itemIds.length).fill(0);

  const { iterations, learningRate, regularization } = config;

  // Gradients are averaged per learner/item (not summed) so the effective
  // step size does not depend on how many responses a learner or item
  // happens to have -- with raw summed gradients, a learner with hundreds
  // of responses would take enormous steps relative to one with a handful,
  // making a single learningRate unstable across datasets of different size.
  const responseCountByLearner = new Float64Array(learnerIds.length);
  const responseCountByItem = new Float64Array(itemIds.length);
  for (const { learnerIndex, itemIndex } of responses) {
    responseCountByLearner[learnerIndex] = (responseCountByLearner[learnerIndex] ?? 0) + 1;
    responseCountByItem[itemIndex] = (responseCountByItem[itemIndex] ?? 0) + 1;
  }

  for (let iter = 0; iter < iterations; iter += 1) {
    const gradTheta = new Float64Array(learnerIds.length);
    const gradA = new Float64Array(itemIds.length);
    const gradB = new Float64Array(itemIds.length);

    for (const { learnerIndex, itemIndex, correct } of responses) {
      const th = theta[learnerIndex] ?? 0;
      const ai = a[itemIndex] ?? 1;
      const bi = b[itemIndex] ?? 0;
      const p = sigmoid(ai * (th - bi));
      const residual = (correct ? 1 : 0) - p; // dLogLik/dz

      gradTheta[learnerIndex] = (gradTheta[learnerIndex] ?? 0) + ai * residual;
      gradB[itemIndex] = (gradB[itemIndex] ?? 0) - ai * residual;
      gradA[itemIndex] = (gradA[itemIndex] ?? 0) + (th - bi) * residual;
    }

    for (let i = 0; i < learnerIds.length; i += 1) {
      const count = responseCountByLearner[i] || 1;
      const grad = (gradTheta[i] ?? 0) / count - regularization * (theta[i] ?? 0);
      theta[i] = (theta[i] ?? 0) + learningRate * grad;
    }
    for (let j = 0; j < itemIds.length; j += 1) {
      const count = responseCountByItem[j] || 1;
      const gradBj = (gradB[j] ?? 0) / count - regularization * (b[j] ?? 0);
      b[j] = (b[j] ?? 0) + learningRate * gradBj;
      const gradAj = (gradA[j] ?? 0) / count - regularization * ((a[j] ?? 1) - 1);
      const nextA = (a[j] ?? 1) + learningRate * gradAj;
      // Discrimination must stay positive; floor it well away from zero.
      a[j] = Math.max(nextA, 0.05);
    }

    // Fix the theta location/scale gauge freedom (see docstring above).
    let mean = 0;
    for (let i = 0; i < theta.length; i += 1) mean += theta[i] ?? 0;
    mean /= theta.length || 1;
    let variance = 0;
    for (let i = 0; i < theta.length; i += 1) variance += ((theta[i] ?? 0) - mean) ** 2;
    variance /= theta.length || 1;
    const std = Math.sqrt(variance);
    if (std > 1e-6) {
      for (let i = 0; i < theta.length; i += 1) {
        theta[i] = ((theta[i] ?? 0) - mean) / std;
      }
      for (let j = 0; j < itemIds.length; j += 1) {
        b[j] = ((b[j] ?? 0) - mean) / std;
        a[j] = (a[j] ?? 1) * std;
      }
    }
  }

  const thetaMap = new Map<string, number>();
  learnerIds.forEach((id, i) => thetaMap.set(id, theta[i] ?? 0));
  const aMap = new Map<string, number>();
  const bMap = new Map<string, number>();
  itemIds.forEach((id, j) => {
    aMap.set(id, a[j] ?? 1);
    bMap.set(id, b[j] ?? 0);
  });

  return { theta: thetaMap, a: aMap, b: bMap };
}

/**
 * Item Response Theory (2-parameter logistic) scoring model. Treats each
 * distinct skillId as one "item": every response event for a skill is an
 * observation of that item, and the model jointly estimates one ability
 * (theta) per learner and one (discrimination, difficulty) pair per skill.
 * Implements the ScoringModel interface so it is interchangeable with BKT.
 */
export class IrtModel implements ScoringModel<IrtFittedModel> {
  public readonly name = 'irt';

  private readonly config: Required<IrtConfig>;

  constructor(config: IrtConfig = {}) {
    this.config = { ...DEFAULT_IRT_CONFIG, ...config };
  }

  fit(events: ResponseEvent[]): IrtFittedModel {
    const learnerIds = [...new Set(events.map((e) => e.learnerId))];
    const itemIds = [...new Set(events.map((e) => e.skillId))];
    const learnerIndex = new Map(learnerIds.map((id, i) => [id, i]));
    const itemIndex = new Map(itemIds.map((id, i) => [id, i]));

    const responses = events.map((e) => ({
      learnerIndex: learnerIndex.get(e.learnerId) ?? 0,
      itemIndex: itemIndex.get(e.skillId) ?? 0,
      correct: e.correct,
    }));

    const responseCounts = new Map<string, number>();
    for (const e of events) {
      responseCounts.set(e.learnerId, (responseCounts.get(e.learnerId) ?? 0) + 1);
    }

    if (learnerIds.length === 0 || itemIds.length === 0) {
      return {
        modelName: 'irt',
        items: [],
        learners: [],
        responses: [],
      };
    }

    const { theta, a, b } = fitJmle(learnerIds, itemIds, responses, this.config);

    const items: IrtItemParams[] = itemIds.map((skillId) => ({
      skillId,
      a: a.get(skillId) ?? 1,
      b: b.get(skillId) ?? 0,
    }));

    const learners: IrtLearnerResult[] = learnerIds.map((learnerId) => ({
      learnerId,
      theta: theta.get(learnerId) ?? 0,
      responseCount: responseCounts.get(learnerId) ?? 0,
    }));

    return {
      modelName: 'irt',
      items,
      learners,
      responses: events.map((e) => ({
        learnerId: e.learnerId,
        skillId: e.skillId,
        correct: e.correct,
      })),
    };
  }

  score(fittedModel: IrtFittedModel): MasteryReport {
    const itemsBySkill = new Map(fittedModel.items.map((item) => [item.skillId, item]));
    const skillIdsByLearner = new Map<string, Set<string>>();
    const responseCountByLearnerSkill = new Map<string, number>();

    for (const response of fittedModel.responses) {
      const set = skillIdsByLearner.get(response.learnerId) ?? new Set<string>();
      set.add(response.skillId);
      skillIdsByLearner.set(response.learnerId, set);

      const key = `${response.learnerId}::${response.skillId}`;
      responseCountByLearnerSkill.set(key, (responseCountByLearnerSkill.get(key) ?? 0) + 1);
    }

    const learners: MasteryLearnerEntry[] = fittedModel.learners.map((learnerResult) => {
      const skillIds = [...(skillIdsByLearner.get(learnerResult.learnerId) ?? [])];
      const skills = skillIds.map((skillId) => {
        const item = itemsBySkill.get(skillId);
        const a = item?.a ?? 1;
        const b = item?.b ?? 0;
        const key = `${learnerResult.learnerId}::${skillId}`;
        return {
          skillId,
          metric: 'ability_theta' as const,
          value: learnerResult.theta,
          responseCount: responseCountByLearnerSkill.get(key) ?? 0,
          details: {
            itemDiscrimination: a,
            itemDifficulty: b,
            predictedProbabilityCorrect: probabilityCorrect(learnerResult.theta, a, b),
          },
        };
      });
      return { learnerId: learnerResult.learnerId, skills };
    });

    return {
      model: 'irt',
      generatedAt: new Date().toISOString(),
      learners,
      meta: { items: fittedModel.items },
    };
  }
}

File: python/src/masterytrace/models/irt.py
"""
Item Response Theory (2-parameter logistic) scoring model. Faithful port
of src/models/irt.ts -- same joint maximum-likelihood estimation via batch
gradient ascent, same L2 regularization, same per-iteration gauge-fixing
(re-centering theta to mean 0 / std 1). The TypeScript original hand-rolls
this with a `Float64Array` and plain loops rather than calling into an
external optimizer (there is no scipy.optimize equivalent in play here to
port to); this port mirrors that with plain Python lists so the algorithm
stays line-for-line auditable against the original rather than being
rewritten in vectorized numpy, which could silently change rounding/
iteration order.
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, List, Optional

from ..core.event_schema import ResponseEvent
from ..core.scoring_model import FittedModel, MasteryLearnerEntry, MasteryReport, MasterySkillEntry


@dataclass(frozen=True)
class IrtItemParams:
    skill_id: str
    a: float
    """Discrimination. Higher means the item separates high/low ability learners more sharply."""
    b: float
    """Difficulty, on the same scale as theta. Higher means a harder skill."""


@dataclass(frozen=True)
class IrtLearnerResult:
    learner_id: str
    theta: float
    """Estimated ability."""
    response_count: int


@dataclass
class IrtConfig:
    iterations: int = 500
    """Number of gradient-ascent iterations to run."""
    learning_rate: float = 0.5
    """Learning rate for the gradient-ascent updates."""
    regularization: float = 0.01
    """
    L2 regularization strength pulling theta and b toward 0 and a toward 1.
    This is what keeps the joint MLE finite for learners/items with a
    perfect (all-correct or all-incorrect) response record, where the
    unregularized likelihood is maximized at +/-infinity.
    """


@dataclass
class IrtResponseTriple:
    learner_id: str
    skill_id: str
    correct: bool


@dataclass
class IrtFittedModel(FittedModel):
    items: List[IrtItemParams] = field(default_factory=list)
    learners: List[IrtLearnerResult] = field(default_factory=list)
    responses: List[IrtResponseTriple] = field(default_factory=list)


def _sigmoid(z: float) -> float:
    """Numerically stable logistic sigmoid."""
    if z >= 0:
        e = math.exp(-z)
        return 1.0 / (1.0 + e)
    e = math.exp(z)
    return e / (1.0 + e)


def probability_correct(theta: float, a: float, b: float) -> float:
    """P(correct) under the 2-parameter logistic model."""
    return _sigmoid(a * (theta - b))


@dataclass
class _JmleResult:
    theta: Dict[str, float]
    a: Dict[str, float]
    b: Dict[str, float]


def _fit_jmle(
    learner_ids: List[str],
    item_ids: List[str],
    responses: List[tuple],
    config: IrtConfig,
) -> _JmleResult:
    """
    Joint maximum-likelihood estimation for the 2PL IRT model via batch
    gradient ascent on the log-likelihood, with a small L2 prior
    (regularizing theta/b toward 0 and a toward 1) that keeps estimates
    finite for learners or skills with an all-correct or all-incorrect
    record.

    The 2PL model is only identified up to an additive shift and
    multiplicative scale of theta (z = a*(theta-b) is unchanged by
    shifting theta and b by the same constant, or by scaling theta/b by s
    while dividing a by s). To pin down a single solution, after every
    iteration the theta distribution is re-centered to mean 0 and
    re-scaled to standard deviation 1, applying the matching inverse
    transform to b and a so every predicted probability is left exactly
    unchanged. This is the standard way JMLE implementations fix the
    person-parameter scale.
    """
    n_learners = len(learner_ids)
    n_items = len(item_ids)

    theta = [0.0] * n_learners
    a = [1.0] * n_items
    b = [0.0] * n_items

    iterations = config.iterations
    learning_rate = config.learning_rate
    regularization = config.regularization

    # Gradients are averaged per learner/item (not summed) so the
    # effective step size does not depend on how many responses a learner
    # or item happens to have -- with raw summed gradients, a learner with
    # hundreds of responses would take enormous steps relative to one with
    # a handful, making a single learning_rate unstable across datasets of
    # different size.
    response_count_by_learner = [0.0] * n_learners
    response_count_by_item = [0.0] * n_items
    for learner_index, item_index, _correct in responses:
        response_count_by_learner[learner_index] += 1
        response_count_by_item[item_index] += 1

    for _ in range(iterations):
        grad_theta = [0.0] * n_learners
        grad_a = [0.0] * n_items
        grad_b = [0.0] * n_items

        for learner_index, item_index, correct in responses:
            th = theta[learner_index]
            ai = a[item_index]
            bi = b[item_index]
            p = _sigmoid(ai * (th - bi))
            residual = (1.0 if correct else 0.0) - p  # dLogLik/dz

            grad_theta[learner_index] += ai * residual
            grad_b[item_index] -= ai * residual
            grad_a[item_index] += (th - bi) * residual

        for i in range(n_learners):
            count = response_count_by_learner[i] or 1.0
            grad = grad_theta[i] / count - regularization * theta[i]
            theta[i] = theta[i] + learning_rate * grad

        for j in range(n_items):
            count = response_count_by_item[j] or 1.0
            grad_bj = grad_b[j] / count - regularization * b[j]
            b[j] = b[j] + learning_rate * grad_bj
            grad_aj = grad_a[j] / count - regularization * (a[j] - 1.0)
            next_a = a[j] + learning_rate * grad_aj
            # Discrimination must stay positive; floor it well away from zero.
            a[j] = max(next_a, 0.05)

        # Fix the theta location/scale gauge freedom (see docstring above).
        mean = sum(theta) / n_learners if n_learners else 0.0
        variance = sum((t - mean) ** 2 for t in theta) / n_learners if n_learners else 0.0
        std = math.sqrt(variance)
        if std > 1e-6:
            theta = [(t - mean) / std for t in theta]
            for j in range(n_items):
                b[j] = (b[j] - mean) / std
                a[j] = a[j] * std

    theta_map = {learner_ids[i]: theta[i] for i in range(n_learners)}
    a_map = {item_ids[j]: a[j] for j in range(n_items)}
    b_map = {item_ids[j]: b[j] for j in range(n_items)}
    return _JmleResult(theta=theta_map, a=a_map, b=b_map)


class IrtModel:
    """
    Item Response Theory (2-parameter logistic) scoring model. Treats
    each distinct skillId as one "item": every response event for a skill
    is an observation of that item, and the model jointly estimates one
    ability (theta) per learner and one (discrimination, difficulty) pair
    per skill. Implements the ScoringModel protocol so it is
    interchangeable with BKT.
    """

    name = "irt"

    def __init__(self, config: Optional[IrtConfig] = None) -> None:
        self.config = config or IrtConfig()

    def fit(self, events: List[ResponseEvent]) -> IrtFittedModel:
        learner_ids: List[str] = []
        seen_learners = set()
        item_ids: List[str] = []
        seen_items = set()
        for e in events:
            if e.learner_id not in seen_learners:
                seen_learners.add(e.learner_id)
                learner_ids.append(e.learner_id)
            if e.skill_id not in seen_items:
                seen_items.add(e.skill_id)
                item_ids.append(e.skill_id)

        learner_index = {lid: i for i, lid in enumerate(learner_ids)}
        item_index = {sid: i for i, sid in enumerate(item_ids)}

        responses = [(learner_index[e.learner_id], item_index[e.skill_id], e.correct) for e in events]

        response_counts: Dict[str, int] = {}
        for e in events:
            response_counts[e.learner_id] = response_counts.get(e.learner_id, 0) + 1

        if not learner_ids or not item_ids:
            return IrtFittedModel(model_name="irt", items=[], learners=[], responses=[])

        fitted = _fit_jmle(learner_ids, item_ids, responses, self.config)

        items = [IrtItemParams(skill_id=sid, a=fitted.a.get(sid, 1.0), b=fitted.b.get(sid, 0.0)) for sid in item_ids]
        learners = [
            IrtLearnerResult(
                learner_id=lid,
                theta=fitted.theta.get(lid, 0.0),
                response_count=response_counts.get(lid, 0),
            )
            for lid in learner_ids
        ]

        return IrtFittedModel(
            model_name="irt",
            items=items,
            learners=learners,
            responses=[IrtResponseTriple(learner_id=e.learner_id, skill_id=e.skill_id, correct=e.correct) for e in events],
        )

    def score(self, fitted_model: IrtFittedModel) -> MasteryReport:
        items_by_skill = {item.skill_id: item for item in fitted_model.items}
        skill_ids_by_learner: Dict[str, List[str]] = {}
        seen_pairs = set()
        response_count_by_learner_skill: Dict[str, int] = {}

        for response in fitted_model.responses:
            key = f"{response.learner_id}::{response.skill_id}"
            pair_key = (response.learner_id, response.skill_id)
            if pair_key not in seen_pairs:
                seen_pairs.add(pair_key)
                skill_ids_by_learner.setdefault(response.learner_id, []).append(response.skill_id)
            response_count_by_learner_skill[key] = response_count_by_learner_skill.get(key, 0) + 1

        learners: List[MasteryLearnerEntry] = []
        for learner_result in fitted_model.learners:
            skill_ids = skill_ids_by_learner.get(learner_result.learner_id, [])
            skills: List[MasterySkillEntry] = []
            for skill_id in skill_ids:
                item = items_by_skill.get(skill_id)
                a = item.a if item else 1.0
                b = item.b if item else 0.0
                key = f"{learner_result.learner_id}::{skill_id}"
                skills.append(
                    MasterySkillEntry(
                        skill_id=skill_id,
                        metric="ability_theta",
                        value=learner_result.theta,
                        response_count=response_count_by_learner_skill.get(key, 0),
                        details={
                            "item_discrimination": a,
                            "item_difficulty": b,
                            "predicted_probability_correct": probability_correct(learner_result.theta, a, b),
                        },
                    )
                )
            learners.append(MasteryLearnerEntry(learner_id=learner_result.learner_id, skills=skills))

        return MasteryReport(
            model="irt",
            generated_at=datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            learners=learners,
            meta={"items": fitted_model.items},
        )
